pg_upgrade check feature

degaurab opened the pull request:
Adds `--skip-target-check` and `--continue-check-on-fatal` features for check mode:


### `--continue-check-on-fatal`
The  flag lets us traverse through the pg_upgrade --check list without
exiting the first error encountered. It allows the user to get a report of 
the entire list of non-upgradable objects in their cluster at one go.

Note: We canot ignore certain checks despite the flag being supplied -
they may be critical to the check process . One such example is
heck_proper_datallowconn(), which ensures that we can connect to user
databases to perform the checks themselves.  Specific checks are
multi-step, like the check to ensure that only the install user is the
sole user in the target database. If we get fatal: "could not determine
the number of users", we can't proceed with the check. Thus, in these
cases, the flag will not respect the flag.

Example:
```
pg_upgrade \
-b /usr/local/5XSTABLE/bin/ \
-B /usr/local/6XSTABLE/bin/ \
-d /data/5XStable/demoDataDir-1 \
-D /data/6XStable/demoDataDir-1 \
-p 15432 \
-P 6000 \
-c \
--continue-check-on-fatal
```

### `--skip-target-check`
The flag allows us to skip target cluster checks performed by pg_upgrade
--check. It lets users find out incompatible objects on the source
cluster without the hassle of setting up a target cluster. For instance,
it is often non-trivial to set up GPDB specific extensions, and users
invariably run into a failure in check_loadable_libraries().

- need to be used along with check flag (-c)
- when you are using the flag we don't need to provide:
  - new-bindir (-B)
  - new-datadir (-D)
  - new-port (-P)
- it can be used along with `--continue-check-on-fatal`

Example:
```
pg_upgrade \
-b /usr/local/5XSTABLE/bin/ \
-d /data/5XStable/demoDataDir-1 \
-p 15432 \
-c \
--continue-check-on-fatal \
--skip-target-check
```

Dependency: https://github.com/greenplum-db/gpdb/pull/13122

degaurab commented:
Need to merge: https://github.com/greenplum-db/gpdb/pull/13122 before approving this PR

bhuvnesh2703 commented on src/bin/pg_upgrade/util.c:
```diff
@@ -178,6 +178,28 @@ check_ok(void)
+
+	if(is_continue_check_on_fatal())
+	{
+		check_error_occured();
```
at some places fatal and other places error, we should make it consistent, may be use *fatal*

bhuvnesh2703 commented on src/bin/pg_upgrade/greenplum/pg_upgrade_greenplum.h:
```diff
@@ -42,28 +42,38 @@ typedef enum {
       --remove-checksum         remove data checksums when creating new cluster\n\
       --add-checksum            add data checksumming to the new cluster\n\
+			--continue-check-on-fatal goes through all pg_upgrade checks, without upgrade\n\
+			--skip-target-check       skip checks all checks for new cluster\n\
```
formatting

bhuvnesh2703 commented on src/bin/pg_upgrade/greenplum/option_gp.c:
```diff
@@ -73,3 +102,26 @@ is_show_progress_mode(void)
+}
+
+bool
+get_check_error(void)
```
error vs fatal (similar to other comment)

bhuvnesh2703 commented on src/bin/pg_upgrade/greenplum/option_gp.c:
```diff
@@ -48,6 +51,32 @@ process_greenplum_option(greenplumOption option, char *option_value)
+			else
+			{
+				pg_log(PG_FATAL,
+					"--continue-check-on-fault: should be used with check mode (-c)\n");
```
typo: continue-check-on-fatal

bhuvnesh2703 commented on src/bin/pg_upgrade/check.c:
```diff
@@ -240,7 +240,9 @@ report_clusters_compatible(void)
 	{
-		pg_log(PG_REPORT, "\n*Clusters are compatible*\n");
+		pg_log(PG_REPORT, (get_check_error() ?
+			"\n*Clusters are NOT compatible*\n" : "\n*Clusters are compatible*\n"));
```
when skip target checks is used, would it be better to print something for the form, that the cluster has objects which cannot be upgrade. Since Clusters are NOT compatible means a comparison with a target cluster.

bhuvnesh2703 commented on src/bin/pg_upgrade/greenplum/pg_upgrade_greenplum.h:
```diff
@@ -42,28 +42,38 @@ typedef enum {
 /* option_gp.c */
 void initialize_greenplum_user_options(void);
-bool process_greenplum_option(greenplumOption option, char *option_value);
+bool process_greenplum_option(greenplumOption option);
```
the implementation seems to have not changed, is this intentional?

bhuvnesh2703 commented on src/bin/pg_upgrade/greenplum/option_gp.c:
```diff
@@ -73,3 +102,26 @@ is_show_progress_mode(void)
+void
+check_error_occured(void)
+{
+	fatal_on_continue_check = true;
```
it's setting the value, so the name should change.

shivzone commented:
Should we considering adding tests with these new flags in `test_gpdb.sh` ?

yaowangm commented:
Just personal thoughts:

The name "--continue-check-on-fatal" is sort of confusing to me. Why need we continue even hitting fatal error? In fact, the option is used only on check mode and to check as much as possible with ignoring errors. Is it correct? If yes, the errors should not be treated as "fatal" in the context. Maybe it's better to use "--continue-check-on-error" or "--ignore-error-of-check".

Anyway it's debatable and up to you.

yaowangm commented on src/bin/pg_upgrade/check.c:
```diff
@@ -368,6 +357,32 @@ check_cluster_versions(void)
+
 	/* Only current PG version is supported as a target */
 	if (GET_MAJOR_VERSION(new_cluster.major_version) != GET_MAJOR_VERSION(PG_VERSION_NUM))
 		pg_fatal("This utility can only upgrade to Greenplum version %s.\n",
```
Not related to your change. The error message is about "Greenplum version %s", but the parameter provided is PG_MAJORVERSION ("12" for GPDB7). Should it be changed to GP_VERSION_NUM or something like "Greenplum version %s based on PG version %s"?

yaowangm commented on src/bin/pg_upgrade/greenplum/option_gp.c:
```diff
@@ -73,3 +104,26 @@ is_show_progress_mode(void)
+}
+
+void
+set_check_fatal_occured(void)
```
Since we have a public var `fatal_on_continue_check`, is it necessary to use wrapper functions? I don't see the necessity unless we have a plan to add more code to the functions in future.

yaowangm commented on src/bin/pg_upgrade/greenplum/option_gp.c:
```diff
@@ -73,3 +104,26 @@ is_show_progress_mode(void)
+}
+
+bool
+get_check_fatal(void)
```
If we decide to use wrapper functions, I suggest to change the name to keep consistency with var name, like `get_severe_error_occurred_oncheck`.

yaowangm commented on src/bin/pg_upgrade/greenplum/option_gp.c:
```diff
@@ -10,14 +10,19 @@ typedef struct {
 } GreenplumUserOpts;
 
 static GreenplumUserOpts greenplum_user_opts;
+static bool fatal_on_continue_check;
```
The name `fatal_on_continue_check` is difficult to be distinguished from `GreenplumUserOpts.continue_check_on_fatal.` Maybe a better name is `check_fatal_occurred` or `severe_error_occurred_oncheck`.

yaowangm commented on src/bin/pg_upgrade/util.c:
```diff
@@ -178,6 +178,28 @@ check_ok(void)
+ * with --continue-check-on-fatal
+ */
+void
+gp_fatal_log(const char *fmt,...)
```
Again, the function name `gp_fatal_log` is sort of confusing because it's actually PG_WARNING when `continue_check_on_fatal` is set. And `gp_fatal_log` is tend to be understood as `pg_log` with `FATAL`. Maybe a better name is `gp_severe_log`?

yaowangm reviewed (commented):
Just some comments about coding style. Please ignore them if they don't make sense to you.

soumyadeep2007 reviewed (commented):
Initial q about the first commit:

Should the continue flag apply only to GP specific checks (`check_gp.c`) or to all checks? Why do we have only certain PG checks respect the flag (like `check_for_isn_and_int8_passing_mismatch`) but not others (like `check_for_jsonb_9_4_usage`)?

degaurab commented:
> Initial q about the first commit:
> 
> Should the continue flag apply only to GP specific checks (`check_gp.c`) or to all checks? Why do we have only certain PG checks respect the flag (like `check_for_isn_and_int8_passing_mismatch`) but not others (like `check_for_jsonb_9_4_usage`)?

as per the discussion we need to skip only the list of checks which are displayed in output. That's a good catch I think I missed it in my search. Will cross-check any other check that I might have missed.

kalensk commented:
> as per the discussion we need to skip only the list of checks which are displayed in output. Will cross-check any other check that I might have missed.

Just to be clear --continue-on-fatal/error needs to execute "all" checks including "both" gpdb specific "and" pg_upgrade specific.

Otherwise this defeats the entire purpose of the feature which is to allow customers to see what occurs they need to fix as early as possible in the upgrade process, and fix them with minimal downtime.

degaurab commented:
> Just to be clear --continue-on-fatal/error needs to execute "all" checks including "both" gpdb specific "and" pg_upgrade specific.
> 
> Otherwise this defeats the entire purpose of the feature which is to allow customers to see what occurs they need to fix as early as possible in the upgrade process, and fix them with minimal downtime.

@kalensk its not including `all` the checks that get executed behind the scenes right now. We discussed as the first take on this feature we are only including the checks which are visible to users.

degaurab commented:
As per discussion we added note in `src/bin/pg_upgrade/README.gpdb` regarding steps for upstream merge, to make sure we are continuing to support new check flags.

shivzone commented:
It would be ideal if we can avoid bailing out for all relevant error cases with the flag `--continue-check-on-fatal`. 
That said, I'm totally fine with the the iterative  approach that we are taking as part of this PR as long as we document this when we are exposing this functionality to end users (eg: gpupgrade), to set us in the path to identify as many inconsistencies as possible.

soumyadeep2007 commented on src/bin/pg_upgrade/pg_upgrade.c:
```diff
@@ -110,7 +110,9 @@ main(int argc, char **argv)
+	if(!is_skip_target_check())
+		adjust_data_dir(&new_cluster);
 
 	setup(argv[0], &live_check);
```
Inside `setup`, we should skip the following which will attempt to start and stop the new postmaster:
```
/* same goes for the new postmaster */
	if (pid_lock_file_exists(new_cluster.pgdata))
```
with a guard: `!is_skip_target_check()`

soumyadeep2007 commented on src/bin/pg_upgrade/pg_upgrade.c:
```diff
@@ -122,7 +124,8 @@ main(int argc, char **argv)
+	 check_cluster_compatibility(live_check);
 
 	/* Set mask based on PGDATA permissions */
 	if (!GetDataDirectoryCreatePerm(new_cluster.pgdata))
```
Needs to be guarded with `!is_skip_target_check()`

soumyadeep2007 commented on src/bin/pg_upgrade/pg_upgrade.c:
```diff
@@ -122,7 +124,8 @@ main(int argc, char **argv)
 
-	check_cluster_compatibility(live_check);
+	if(!is_skip_target_check())
+	 check_cluster_compatibility(live_check);
```
Inside `check_cluster_compatibility` we have to protect the following w/ `!is_skip_target_check`:
```
get_control_data(&new_cluster, false);
check_control_data(&old_cluster.controldata, &new_cluster.controldata);
```
and also:
```
if (live_check && old_cluster.port == new_cluster.port)
		pg_fatal("When checking a live server, "
				 "the old and new port numbers must be different.\n");
```

Otherwise, we are sure to run into a fatal error, if the target cluster is not installed.

soumyadeep2007 replied:
Further below, inside `check_and_dump_old_cluster`, we have to guard the following as well:

Inside `init_tablespaces`:
```
set_tablespace_directory_suffix(&new_cluster);

if (os_info.num_old_tablespaces > 0 &&
	strcmp(old_cluster.tablespace_suffix, new_cluster.tablespace_suffix) == 0)
	pg_fatal("Cannot upgrade to/from the same system catalog version when\n"
				 "using tablespaces.\n");
```

soumyadeep2007 reviewed (commented):
There are a few PG user facing checks that currently don't acknowledge
the continue flag:
1. `check_for_prepared_transactions`
2. `check_for_pg_role_prefix`
3. `check_for_jsonb_9_4_usage`
4. `check_new_cluster_is_empty`
5. `check_locale_and_encoding`
6. `check_is_install_user`
7. `check_loadable_libraries`

They need to acknowledge the flag. **Doing so will remove the need to make the
distinction of a user-facing vs non-user-facing check.** See the proposed commit
message below for an updated justification for checks that cannot respect the
flag.


Please change the commit message body for the --continue-check-on-fatal to the
following:

```
The --continue-check-on-fatal flag allows us to plough through the
pg_upgrade --check list without exiting on the first error encountered.
This allows the user to get a report of the entire list of
non-upgradable objects in their cluster at one go.

Note that there are certain checks that cannot be ignored in spite of
the flag being supplied - they may be critical to the check process
itself. One such example is check_proper_datallowconn(), which ensures
that we can connect to user databases to perform the checks themselves.
Certain checks are multi-step, like the check to ensure that only the
install user is the sole user in the target database. If we get the
fatal: "could not determine the number of users", we can't really
proceed with the check. Thus, in these cases the flag will not be
respected.

Example usage:

pg_upgrade \
-b /usr/local/5XSTABLE/bin/ \
-B /usr/local/6XSTABLE/bin/ \
-d /data/5XStable/demoDataDir-1 \
-D /data/6XStable/demoDataDir-1 \
-p 15432 \
-P 6000 \
--check \
--continue-check-on-fatal
```

Please change the commit message body for the --skip-target-check flag to the
following:

```
The --skip-target-check flag allows us to skip target cluster checks
performed by pg_upgrade --check. This can let users find out
incompatible objects on the source cluster without the hassle of
setting up a target cluster. For instance, it is often non-trivial to
set up GPDB specific extensions and users invariably run into a failure
in check_loadable_libraries().

Note: when this flag is specified we no longer need to pass in:
  - new-bindir (-B)
  - new-datadir (-D)
  - new-port (-P)

Typically, it will be mostly used along with the
--continue-check-on-fatal flag, which was introduced in an earlier
commit.

Example usage:

pg_upgrade \
-b /usr/local/5XSTABLE/bin/ \
-d /data/5XStable/demoDataDir-1 \
-p 15432 \
--check \
--skip-target-check
```

At the end, please update the PR summary to include the new commit
messages as well. Thanks for working on this! Let me know when you
have addressed the feedback and I will approve the PR.

Backport notes: 6X pg_upgrade code will have checks that are not
present in the master branch and vice-versa. Also, 6X has the
--old-gp-dbid and --new-gp-dbid flags, so please include them in the
commit message example.

degaurab commented:
thanks @soumyadeep2007  for the feedback 

> There are a few PG user facing checks that currently don't acknowledge the continue flag:
> 
> 1. `check_for_prepared_transactions`
> 2. `check_for_pg_role_prefix`
> 3. `check_for_jsonb_9_4_usage`
> 4. `check_new_cluster_is_empty`
> 5. `check_locale_and_encoding`
> 6. `check_is_install_user`
> 7. `check_loadable_libraries`
> 

- [X] All the checks are updated to honor --continue-check-on-fatal

> Please change the commit message body for the --continue-check-on-fatal to the following:

- [X] Commit message updated

> Please change the commit message body for the --skip-target-check flag to the following:

- [X] Commit message updated

> At the end, please update the PR summary to include the new commit messages as well. Thanks for working on this! Let me know when you have addressed the feedback and I will approve the PR.

- [X] PR Summary updated

> Backport notes: 6X pg_upgrade code will have checks that are not present in the master branch and vice-versa. Also, 6X has the --old-gp-dbid and --new-gp-dbid flags, so please include them in the commit message example in the backport commits.

Will make sure to include them in the commit message. As it will not be a cherry-pick commit for back-port on 6X_STABLE

soumyadeep2007 reviewed (approved):
LGTM! There are 2 typos in the 1st commit message. And do actually cherry-pick the changes to 6X (it won't be clean but that is fine)

degaurab commented:
we should make sure to merge https://github.com/greenplum-db/gpdb/pull/13122
before merging this PR
